### ISSUE TITLE
Ideas for further proxymodule testing

### ISSUE THREAD

**Issue by cro** [saltstack/salt#39605]
This is a tracking ticket for discussing how proxymodules can be meaningfully tested in the absence of real hardware to test against.

**Comment by mirceaulinic** [comment]
Hi @cro, continuing your idea from https://github.com/saltstack/salt/pull/39575#issuecomment-282044077:

> The challenge there is that all the proxymodules would need to have real devices on the other end or be so heavily mocked that I'm not sure the testing would really tell us anything interesting. 

I agree - there's no point in mocking the tests.
I may have an idea, although it won't be suitable for all the proxy types: we could spawn some Vagrant boxes. For example, for the NAPALM proxy and modules, https://atlas.hashicorp.com/juniper/boxes/ffp-12.1X47-D15.4-packetmode would be good enough to test the basic functionality. What do you think about that?

**Comment by alexbleotu** [comment]
@cro I agree that having end-to-end tests would be ideal but as these states would have side effects, the tests would end up being dependent on one another (the command is the same the result is different based on the existing setup, and to change the existing setup you would probably use states).

Another approach would be to to simply mock the external library (pyvmomi in this case) and check the results based on various state files, pillars, existing configurations (mocked). That would make the test themselves quite difficult to write and quite big (I don't know how many people would write them) and they would probably cover initial setup only, but at least that would give some certainty that the most common use-cases and oll of the internals keep on working.

In my view, the biggest benefit would be that it would provide working examples of configurations people can copy and addapt.
